frontend/morpheus/store/reducers.js

```diff
@@ -0,0 +1,224 @@
+  }
+};
+
+const buildOfficeState = state => {
+  const { rooms, usersInRoom, officeFilter } = state;
+
+  let office = rooms.map(room => ({
+    id: room.id,
+    name: room.name,
+    users: usersInRoom.filter(u => u.room === room.id).map(u => u.user)
+  }));
+
+  if (officeFilter.onlyFullRoom) {
+    office = office.filter(o => o.users.length > 0);
+  }
+  if (officeFilter.search) {
+    const search = officeFilter.search.toLowerCase();
+    office = office.filter(o => o.name.toLowerCase().includes(search));
+  }
+
+  return {
+    ...state,
+    office
+  };
+};
+
+const buildUsersState = state => {
+  const { usersInRoom, usersFilter, rooms } = state;
+
+  let users = usersInRoom.map(u => {
+    const room = rooms.find(r => r.id === u.room);
+
+    return {
+      id: u.user.id,
+      name: u.user.name,
+      avatar: u.user.imageUrl,
+      inMeet: !!u.user.inMeet,
+      roomId: room ? room.id : "",
+      roomName: room ? room.name : ""
+    };
+  });
+
+  if (usersFilter.search) {
+    const search = usersFilter.search.toLowerCase();
+    users = users.filter(u => u.name.toLowerCase().includes(search));
+  }
+
+  return {
+    ...state,
+    users
+  };
+};
+
+const buildInMeetState = (state, action, inMeet) => {
+  const { id } = action.user;
+  const currentUser = { ...state.currentUser };
+
+  if (currentUser.id === id) {
+    currentUser.inMeet = inMeet;
+  }
+
+  const usersInRoom = state.usersInRoom.map(item => {
+    if (item.user.id === id) {
+      return {
+        ...item,
+        user: {
+          ...item.user,
+          inMeet
+        }
+      };
+    }
+
+    return item;
+  });
+
+  return buildUsersState(
+    buildOfficeState({
+      ...state,
+      currentUser,
+      usersInRoom
+    })
+  );
+};
+
+const reducers = (state = initialState, action) => {
+  switch (action.type) {
+    case SET_CURRENT_USER:
+      return {
+        ...state,
+        currentUser: action.user
+      };
+    case SET_CURRENT_ROOM:
+      return {
+        ...state,
+        currentRoom: action.room
+      };
+    case ADD_ROOMS:
+      return buildOfficeState({
+        ...state,
+        rooms: action.rooms
+      });
+    case CHANGE_OFFICE_FILTER:
+      return buildOfficeState({
+        ...state,
+        officeFilter: {
+          ...state.officeFilter,
+          [action.key]: action.value
+        }
+      });
+    case SYNC_OFFICE:
+      return buildUsersState(
+        buildOfficeState({
+          ...state,
+          usersInRoom: Object.values(action.usersInRoom)
+        })
+      );
+    case CHANGE_USERS_FILTER:
+      return buildUsersState({
+        ...state,
+        usersFilter: {
+          ...state.usersFilter,
+          [action.key]: action.value
+        }
+      });
+    case CHANGE_SETTINGS:
+      return {
+        ...state,
+        settings: {
+          ...state.settings,
+          [action.key]: action.value
+        }
+      };
+    case ADD_USER: {
+      const index = state.usersInRoom.findIndex(
```

Comment:
My suggestion is use a map instead array to store users per room. It's reduce complexity from o(n) to o(1) and simplify add / remove users.

```
const usersInRoom = {}
usersInRoom[action.user.id] = { user: action.user, room: roomId }  } 
Object.values(usersInRoom) // [{user: ...}, {user: ...}]
...
```